lib/doorkeeper/oauth/token_introspection.rb

```diff
@@ -149,9 +149,9 @@ def failure_response
       #
       def active?
         if authorized_client
-          valid_token? && token_introspection_allowed?(authorized_client.application)
+          valid_token? && token_introspection_allowed?
         else
-          valid_token? && token_introspection_allowed?(authorized_token&.application)
```

Comment:
I'm not agree the using of `authorized_token&.application` as the `client` send to introspection config block. `client` param is should only used for `authorized_client`, not `application` of `authorized_token`.
Because at here, the request is authorized by authorized_token method, so the request authorized by authorized_token method will accidentally read the config of client_credential method (for example: token.application == authorized_client)

Reply:
Hi @linhdangduy . Thanks for your work ❤️ 

Actually this check was introduced for the one main reason: previous implementation allowed to introspect any token if you are authorizing using bearer token (without any check). So instead of client credentials that scopes introspected token by application everybody can just use bearer token for auth and see *ANY* token.

So I decided to add the same guard as for client credentials auth - validate if app of the token been introspected is the same as bearer token one. If developer need other rules - he can implement it himself using configuration option.

Im just thinking about the best and security solution by default..

Reply:
@nbulaj For above security reason by default, we can _set that check at default configuration_.
Because of the flexibility this configuration provides, misuse of configuration can lead to security problem. I think it's better:
- One parameter represents only one role.
- Adding to doorkeeper doc: examples for properly use of config.

Should I write the the config as the following:
```ruby
# config.rb
option :allow_token_introspection,
       default: (lambda do |token, authorized_client, authorized_token|
         if !token.application
           true  # public token
         elsif authorized_token
           # by default: authorized_token and introspected token need to belong to one application
           authorized_token.application == token.application
         elsif authorized_client
           authorized_client == token.application
         end
       end)  
```

Reply:
Yep, looks good for me. his is exactly what I meant :)